lib/graphql/batch.rb

```diff
@@ -1,6 +1,48 @@
 require "graphql"
+
+module GraphQL
+  module Batch
+    module_function
+
+    def execute(container)
+      unless container.is_a?(QueryContainer)
+        return container
+      end
+      if container.completed?
+        raise container.error if container.error
+        return container.result
+      end
+
+      batched_queries = Hash.new{ |hash, key| hash[key] = [] }
+      register_queries = lambda do |query_container|
+        query_container.each_query do |query|
+          batched_queries[query.group_key] << query
+        end
+      end
+      register_queries.call(container)
+
+      until batched_queries.empty?
+        queries = batched_queries.shift.last
+        begin
+          queries.first.class.execute(queries)
+        rescue => err
+          queries.each do |query|
+            query.complete(error: err) unless query.completed?
+          end
+        end
+        queries.each do |query|
+          register_queries.call(query)
```

Comment:
I'm planning on having a global (thread local) executor where queries will be registered so it doesn't have to be tracked on the Query object.  So far I have https://github.com/shopify/graphql-batch/compare/promise which uses promise.rb for promises, but I want to move the query registering out of Query.new.  I think I'll end up changing the API to be `CustomLoader.for(*group_args).load(key)` where `.for(*group_args)` will create and register a loader if one doesn't exist on the executor and `.load(key)` will add a key to the loader and return a Promise for that key.